### ISSUE TITLE
some thoughts and ideas about crd lifecycle

### ISSUE THREAD

**Issue by tyriis** [controlplaneio-fluxcd/flux-operator#84]
Hello @stefanprodan

first, thanks a lot for your passion and energy and work youi put into the flux eco system. Flux has changed my life, from pain and chaos with kubernetes, to a shiny great world full of advanced automations and gitops.

I really like the idea of the flux-operator I expect the real benefit to be in a multi tenant cluster where namespace owner can not delete flux by accident, and if so it will be reconciled :D

I am playing arround with it and encountred some things that I would like to understand (if they are intended) (I only use 1 flux instance currently).

When i bootstrap from the operator and the flux instance, everything spins up as expected.
When i delete the instance by accident or intended (cause I want to test out stuff :D)

The operator starts pruning the flux-system, (well okay what would I expect if I delete the instance), problem with this, it start destroy in alphabetic order, ending up in some resources left over as flux-system gest killed before (maybe because I reconcile operator and instance with flux aswel?).

Sad story some kustomizaztions are still present while others are gone, so far fine, but it is not possible to re-apply the instance as for some reason the crds get uninstalled (but cant as they are items left over and as thy have a flux finalizer they need manual help now :(

Just a thought:

lets setup all flux crds with the operator (why would I install the operator if I will never spin up any instance?) pretty sure at some point I need them. To not deploy all crds (I personaly have no use for the image-automation part) I could imagine a controller based config like in currently in the instance (if an instance try to use the image-automation but it is nor crd installed by operator it should throw some error to clarify that the flux admin need to activate it on the controller level first.
With this approach crds can be lifecycled with the operator. Anyway you will always have some specific version of crds in the cluster so in the worst case the any of the flux-controllers will throw errors when deployed in a lower verison as the new crds, in case new features are used.

I have another maybe only personal issue with how the root flux kustomization is defined, in case operator is deploying flux-crds initialy it would be possible to reference the root kustomization and source with the already well known flux resources and reference them.

Maybe a circular loop is just not intended but in any case I am unhappy that flux instance is deleting crds (would be great to prevent it!) (also prune(false) behaviour in kustomizations is holding back a deletion of all resources.

example:

```yaml
apiVersion: fluxcd.controlplane.io/v1
kind: FluxInstance
metadata:
  name: flux
spec:
  uninstall:
  	crds: Keep # default to Delete?
```

Maybe one last thing (when adopting an existing cluster flux-operator is replacing resources instead of adopting, would be great to not destroy an existing system durring migration :D

Thanks in advance for your feedback <3

**Comment by stefanprodan** [comment]
> I really like the idea of the flux-operator I expect the real benefit to be in a multi tenant cluster where namespace owner can not delete flux by accident

Flux Operator is meant to be installed by a cluster admin. It can only be only one FluxInstance per cluster, created in the same namespace where the operator is installed. No tenant should have access to that namespace, so the issue you described does apply here.

> The operator starts pruning the flux-system

This is not the case, the operator does not delete the flux-system namespace when you delete the `FluxInstance`. Deleting the namespace where the operator is installed is not something a cluster admin should do. If you don't trust your admins, then take away their power to delete the flux-system namespace with RBAC.

> when adopting an existing cluster flux-operator is replacing resources instead of adopting

This is not the case, the operator doesn't perform any delete operations, just sever-side apply.

**Comment by tyriis** [comment]
maybe i got you wrong, in my case it is deleting the flux-system content, together with all flux kustomizations as it tries to delete the CRD definition, reaching the point where flux-controllers are decomissioned the procedure stops cause of no longer available finalizers.

here a few screens of the behaviour:

![Screenshot_20240812_221401](https://github.com/user-attachments/assets/f6c7ee38-9f7b-4c77-a73f-e5b3d1f427c3)
![Screenshot_20240812_221737](https://github.com/user-attachments/assets/1943a75e-4996-4eb1-9aac-ec427e627792)
![Screenshot_20240812_222731](https://github.com/user-attachments/assets/d3ffd84b-5f69-451a-b751-caf6816e5844)

**Comment by stefanprodan** [comment]
That’s correct, it should remove Flux but not the namespace. It’s the same as running `flux uninstall`. Are you saying that CRDs got stuck instead of being removed?

**Comment by tyriis** [comment]
Yes exactly they have finalizers but are unable to fulfill as flux controllers are removed before all kustomizations are removed. Any chance to get a crd keep option? I really don't want to have a destroy all option build-in. Uninstalling flux does not mean uninstalling everything flux is managing (at least for me).

As you wrote, operator is only supposed to create 1 instance, then there is no real reason to not install flux crds with operator (or is there a use case where operator is installed but flux is not?)

**Comment by stefanprodan** [comment]
Uninstall does not remove any managed workloads from the cluster, it just removes the Flux definitions. All CRDs must be wiped otherwise you can't rollback to an older version. This is how `flux uninstall` works since I first wrote it years ago. Can you please paste here the logs of uninstall if you see any errors, no CRD should be stuck.

**Comment by stefanprodan** [comment]
Ok I see in your printscreen the error, it's a conflict, probably due to etcd slowness. I have opened a PR to fix this #85

**Comment by tyriis** [comment]
Will check if it solves all effects.

If I got you right, it is not expected that managed helm charts are uninstalled (prune:false is set on kustomization level). But exactly this happen to me (have replayed it a few times).

**Comment by stefanprodan** [comment]
It is expected for none of the managed workloads to be touched, if you see anything like that please let me know, it would be a major bug.

**Comment by stefanprodan** [comment]
The only way Helm release are being deleted is when Flux Operator does not managed helm-controller, or if you have more than one helm-controller installed by other means on the cluster.

**Comment by tyriis** [comment]
I can confirm that in version 0.8 it is now behaving as you have described (same behaviour as flux uninstall)

I have tested the following szenarios:
- existing cluster: flux uninstall, operator -> instance -> delete flux instance
- existing cluster: operator -> instance adopt (no self reconsilce) -> delete flux instance
- existing cluster: operator -> instance adopt (with self reconcile) -> delete flux instance

**Comment by stefanprodan** [comment]
Great thanks for testing it. I hope now there no more issues with the managed workloads, besides Flux no other deployments get removed?

**Comment by tyriis** [comment]
> Great thanks for testing it. I hope now there are no more issues with the managed workloads, besides Flux, no other deployment gets removed?

yes it is working now, thank you

**Comment by tyriis** [comment]
thought about it again, because I wanted to understand why it has started deleting stuff (and answer me the question, can this happen again?)

I think it is because of not removing finalizer but deleting the crd item, flux-controller kicks in and executes the deletion (and as prune was true, it started purging my apps).

I think it is fixed now as you wait for the finalizers to be removed before deleting the crd content.

Anyway, if the idea is to just install 1 flux instance in 1 cluster, there is no real advantage of installing crds with the instance instead of the operator. Then you would be able to directly configure flux kustomizations and repository for the starting point in the already known flux fashion, instead of re-modeling a subset that need kustomize patches for adoptions.

**Comment by stefanprodan** [comment]
> I think it is because of not removing finalizer but deleting the crd item, flux-controller kicks in and executes the deletion (and as prune was true, it started purging my apps).

This can't happen even if the finalizers were stuck because there is no Flux controller running to perform the app deletion. The operator first deletes the Kubernetes Deployments of kustomize-controller, helm-controller etc, and only then it moves to the custom resources and finally CRDs. Going to ask you again, when you observed the app deletion, where there multiple Flux installations on your cluster?

**Comment by tyriis** [comment]
No i have had only a single flux-controller set (source, helm, kustomize, notification) in flux-system running. I can reproduce the behaviour in the old version. in 0.8.0 version it is not happening. (If my guess is wrong, what can it be else?)

**Comment by stefanprodan** [comment]
@tyriis I've found the condition when this happens, the fix is in #90
